osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyUserRankCounts.cs

```diff
@@ -0,0 +1,161 @@
+        public int RulesetId { get; set; }
+
+        [Option(CommandOptionType.SingleOrNoValue, Template = "-v|--verbose", Description = "Output verbose information on processing.")]
+        public bool Verbose { get; set; }
+
+        [Option(CommandOptionType.SingleOrNoValue, Template = "--dry-run")]
+        public bool DryRun { get; set; }
+
+        public async Task<int> OnExecuteAsync(CancellationToken cancellationToken)
+        {
+            LegacyDatabaseHelper.RulesetDatabaseInfo databaseInfo = LegacyDatabaseHelper.GetRulesetSpecifics(RulesetId);
+
+            Console.WriteLine($"Running for ruleset {RulesetId}");
+
+            if (DryRun)
+                Console.WriteLine("RUNNING IN DRY RUN MODE.");
+
+            if (!string.IsNullOrEmpty(CustomQuery))
+                CustomQuery = $" AND {CustomQuery}";
+
+            using (var db = await DatabaseAccess.GetConnectionAsync(cancellationToken))
+            {
+                Console.WriteLine("Fetching all users...");
+                uint[] userIds = (await db.QueryAsync<uint>($"SELECT `user_id` FROM {databaseInfo.UserStatsTable} WHERE 1 = 1 {CustomQuery}")).ToArray();
+                Console.WriteLine($"Fetched {userIds.Length} users");
+
+                int processedUsers = 0;
+
+                foreach (uint userId in userIds)
+                {
+                    await processUser(db, userId, cancellationToken);
+
+                    if (++processedUsers % 1000 == 0)
+                        Console.WriteLine($"Processed {processedUsers} of {userIds.Length} users");
+                }
+            }
+
+            return 0;
+        }
+
+        private async Task processUser(MySqlConnection db, uint userId, CancellationToken cancellationToken)
+        {
+            using (var transaction = await db.BeginTransactionAsync(IsolationLevel.ReadCommitted, cancellationToken))
+            {
+                var parameters = new
+                {
+                    userId,
+                    rulesetId = RulesetId,
+                };
+
+                IEnumerable<SoloScore> scores = await db.QueryAsync<SoloScore>(new CommandDefinition(
+                    "SELECT * FROM scores WHERE preserve = 1 AND ranked = 1 AND user_id = @userId AND ruleset_id = @rulesetId",
+                    parameters, cancellationToken: cancellationToken, transaction: transaction));
+
+                var counts = new Dictionary<ScoreRank, uint>
+                {
+                    { ScoreRank.A, 0 },
+                    { ScoreRank.S, 0 },
+                    { ScoreRank.SH, 0 },
+                    { ScoreRank.X, 0 },
+                    { ScoreRank.XH, 0 },
+                };
+
+                if (Verbose)
+                    Console.WriteLine($"Processing user {userId} ({scores.Count()} scores)..");
+
+                IEnumerable<SoloScore?> maxScoresByBeatmap = scores.GroupBy(s => s.beatmap_id).Select(g => g.MaxBy(s => s.total_score));
```

Comment:
The processor also orders the scores by ID descending in case of a total score tie:

https://github.com/ppy/osu-queue-score-statistics/blob/fc43f221bbc46e547346ec7827dbb92b1995263c/osu.Server.Queues.ScoreStatisticsProcessor/Helpers/DatabaseHelper.cs#L182

Not sure this is gonna matter in practice, but there's a chance it might, probably higher than I think. Not sure what this `.MaxBy()` is gonna do in such a case. Wouldn't be surprised if it didn't do the right thing (and by "right" I mean "the same thing that the rank count processor does").

Reply:
That's a really weird choice (to use the most recent score) but I guess it's kinda required in this case for some kind of sanity so I'll just go with it.